# Step by step guide to generate multimapping datasets (FASTQ and BAM)

# General outline
1. Generating FASTQ files containing fragmented reads from the MANE transcriptome v1.4 in silico (to mimic mRNA reads) using Python script.
2. Aligning these reads to the reduced MANE transcriptome using bowtie (using different settings).
![alt text](https://github.com/guydoshlab/Github_figures/readgenerator_fullcov.png)
3. Generate multimapper identifier files (mm_id.BEDGRAPH).
4. Load multimapper identifier files into IGV for visualization of high frequency multimapping regions. 


# Creating a conda environment and installing packages
## Generating the conda environment (only once)
```unix
conda create --name multimap
```
## Activate the enviroment (each time)
```unix
conda activate multimap
```
## Install packages (only once)
https://anaconda.org/bioconda/bowtie

https://anaconda.org/bioconda/fastqc

https://anaconda.org/bioconda/multiqc

https://anaconda.org/bioconda/samtools

```unix
conda install bioconda::bowtie
```
```unix
conda install bioconda::fastqc
```
```unix
conda install bioconda::multiqc
``` 
```unix
conda install bioconda::samtools
``` 

# 1. Generating reads from the transcriptome using Python script
## Download the MANE transcriptome
Download transcriptome FASTA longnames file [here](https://github.com/guydoshlab/ribofootPrinter2.0-beta/tree/main/preparation/MANE_v1.4_Preparation/output_files) (unzip before use). 
We use this version as it contains a single isoform with highest biological relevance.

Run the Python script (readgenerator_fullcov.py) to generate reads to obtain full coverage of each transcript with desired readlength. The output files are FASTQ files. 


# 2. Alignment against MANE transcriptome using bowtie
Different settings were used to determine the effect of footprint size (i.e. read length) on multimapping events. The number of mismatches can be changed by adjusting -v. The -S setting outputs a SAM file. It is important to include the -k 2 setting which allows bowtie to report up to 2 alignments. This is needed to identify multimapping events. 

For more information on bowtie: https://bowtie-bio.sourceforge.net/manual.shtml

A bowtie indexed transcriptome is required for this step. This can be downloaded from the MANE_v1.4_Preparation Github (6 files with ebwt file extension, unzip before use). 

```unix
mkdir -p nomatch
REF=path/to/bowtie_indexed/MANE_transcriptome/MANE_v1.4

for file in *.fq
do
	echo $file
	bowtie -v 0 -y -S -p 12 -k 2 --best --un ./nomatch/${file%.fq}_nomatch.fq -x $REF ./$file ./${file%.fq}.sam	
done
```

Following alignment, bowtie will output information about read alignment. The output from mapping the MANE_100.fq (FASTQ file containing reads with 100 bp lengths) to the reduced transcriptome with -v 0 and -k 2 settings is: 
```unix
# reads processed: 69364686
# reads with at least one alignment: 69364686 (100.00%)
# reads that failed to align: 0 (0.00%)
Reported 71484435 alignments
```
Substracting the "reads with at least one alignment" from "reported alignments" gives the number of multimapped reads (this is possible because we only allow mapping to one other site by using the -k 2 setting). The percentage multimapped reads can be calculated by dividing the number of multimapped reads by "reads processed".
![alt text](https://github.com/kyrakerkhofs/multimapper/blob/main/bowtie_multimapping.png)

# 3. Prepare multimapper identifier files (mm_id.BEGRAPH)
Multimapper identifier files are generated from multimapped reads generated from alignment of transcriptome-generated reads (FASTQ) against the transcriptome (similar as above):
![alt text](https://github.com/kyrakerkhofs/multimapper/blob/main/mm_id.png)

## a. Align transcriptome-derived full coverage FASTQ files to reduced transcriptome (FASTA)
The FASTQ files generated by using the Python script above are used for alignment to the reduced transcriptome with different settings. 
This script includes the parameter -m 1 (instead of -k 2). This instructs bowtie to only report unique alignments. This allows us to divide the reads between uniquely mapped reads (exported as a SAM file) and multimapped reads (exported as a nomatch FASTQ file). The nomatch FASTQ file is the required file for further analysis.

```unix
mkdir -p nomatch
REF=path/to/bowtie_indexed/MANE_transcriptome/MANE_v1.4

for file in *.fq
do
	echo $file
	bowtie -v 0 -y -S -p 12 -m 1 --best --norc --un ./nomatch/${file%.fq}_nomatch.fq -x $REF ./$file ./${file%.fq}.sam	
done
```

Once the FASTQ file containing multimapped reads has been obtained, they are realigned to the reduced transcriptome with bowtie settings allowing multiple multimapped reads (-k 100000). This will output a SAM file containing information on exclusively multimapped reads. 

```unix
cd nomatch
REF=path/to/bowtie_indexed/MANE_transcriptome/MANE_v1.4

for file in *.fq
do
	echo $file
	bowtie -v 0 -y -S -p 12 -k 100000 --best --norc -x $REF ./$file ./${file%.fq}.sam	
done
```
## b. Convert multimapped reads containing SAM file into IGV compatible BEDGRAPH
Convert SAM files to BAM files:
```unix
for file in *.sam
do
	samtools sort -o ${file%.sam}.bam $file
done
```

Convert BAM file into BEDGRAPH file:
```unix
for file in *.bam
do
	echo $file
	bedtools genomecov -ibam $file -bg -5 > ${file%.bam}.bedgraph	
done
```
If desired, 3'-end mapped multimapper identifier files can be generated by replacing -bg -5 with -bg 3. If coverage is desired, see MANE_v1.4_IGV Github page. 

# 4. Upload BEDGRAPH files into IGV
See MANE_v1.4_IGV Github page for more information.
